openwisp_users/backends.py

```diff
@@ -0,0 +1,35 @@
+from django.contrib.auth import get_user_model
+from django.contrib.auth.backends import ModelBackend
+from django.core.exceptions import ValidationError
+from django.db.models import Q
+
+User = get_user_model()
+
+
+class UserAuthenticationBackend(ModelBackend):
+    def authenticate(self, request, username=None, password=None, **kwargs):
+        users = User.objects.filter(
+            Q(email=username) | Q(phone_number=username) | Q(username=username)
+        )
+        # can not use users.first() because it orders the queryset before
+        # returning the first object which is not what we want
+        user = users[0] if users else None
```

Comment:
why aren't you calling `get_user` here?
We can't have two methods doing almost the same thing but differently and one method is not even used in openwisp-users.

What I asked is to get the common logic to filter the user into the `get_user` method, which would be called here and also in openwisp-radius (while openwisp-radius will need to add some more filtering if needed).

Reply:
@nemesisdesign that [django doc](https://docs.djangoproject.com/en/3.1/ref/models/querysets/#first) explains what i said in the comment above. `.first()` orders the queryset before returning the first value. `If the QuerySet has no ordering defined, then the queryset is automatically ordered by the primary key`. So using `.first()` is a problem because the first element might not be element we wanted in this case. I experienced such behaviors during testing.